runtime/reconcilers/model.go

```diff
@@ -644,6 +659,94 @@ func (r *ModelReconciler) acquireExecutor(ctx context.Context, opts *drivers.Mod
 	return executorName, e, release, nil
 }
 
+func (r *ModelReconciler) acquireExecutorWithStage(ctx context.Context, opts *drivers.ModelExecutorOptions) (string, drivers.ModelExecutor, func(), error) {
+	// we want to determine whether stage connector and output connector are compatible
+	// so we get executor but do not execute it since we need to use result of stage 1
+	//
+	// Build model option for stage 2
+	stage2Opts := &drivers.ModelExecutorOptions{
+		InputConnector:   opts.StageConnector,
+		InputProperties:  opts.StageProperties,
+		OutputConnector:  opts.OutputConnector,
+		OutputProperties: opts.OutputProperties,
+	}
+	_, _, tempRel, err := r.acquireExecutor(ctx, stage2Opts)
+	if err != nil {
+		return "", nil, nil, err
+	}
+	tempRel()
+
+	// Build model options for stage 1
+	// set stage props as outputprops
+	stage1Opts := &drivers.ModelExecutorOptions{
+		Env:              opts.Env,
+		ModelName:        opts.ModelName,
+		InputConnector:   opts.InputConnector,
+		InputProperties:  opts.InputProperties,
+		OutputConnector:  opts.StageConnector,
+		OutputProperties: opts.StageProperties,
+	}
+	_, executor, rel, err := r.acquireExecutor(ctx, stage1Opts)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	// execute stage 1
+	res, err := executor.Execute(ctx)
+	if err != nil {
+		rel()
+		return "", nil, nil, err
+	}
+	rel()
+
+	rc, rr, err := r.C.AcquireConn(ctx, res.Connector)
+	if err != nil {
+		return "", nil, nil, err
+	}
+	defer rr()
+
+	// Build model option for stage 2
+	// Use result's connector and result properties as input connector
+	// Typically the result connector will be same as stage connector
+	// but the properties can change
+	stage2Opts = &drivers.ModelExecutorOptions{
+		Env:              opts.Env,
+		ModelName:        opts.ModelName,
+		InputConnector:   res.Connector,
+		InputProperties:  res.Properties,
+		OutputConnector:  opts.OutputConnector,
+		OutputProperties: opts.OutputProperties,
+		Incremental:      opts.Incremental,
+		IncrementalRun:   opts.IncrementalRun,
+		PreviousResult:   opts.PreviousResult,
+	}
+	name, executor, rel, err := r.acquireExecutor(ctx, stage2Opts)
+	if err != nil {
+		// cleanup stage1 result data
+		if mm, ok := rc.AsModelManager(r.C.InstanceID); ok {
+			return "", nil, nil, errors.Join(err, mm.Delete(ctx, res))
+		}
+		return "", nil, nil, err
+	}
+	// the final cleanup should also cleanup stage1 result data
+	releaseWithCleanUp := func() {
+		rel()
+		rc, rr, err := r.C.AcquireConn(context.Background(), res.Connector)
+		if err != nil {
+			return
+		}
+		defer rr()
+		if mm, ok := rc.AsModelManager(r.C.InstanceID); ok {
+			// May block reconcile. May be add a timeout and run in background ?
+			err = mm.Delete(context.Background(), res)
+			if err != nil {
+				r.C.Logger.Warn("failed to clean up stage output", zap.Error(err))
+			}
+		}
+	}
```

Comment:
Yeah the use of `context.Background()` in a reconciler is quite dangerous and something we need to avoid because the controller assumes that reconcilers are "cooperative" i.e. when a `ctx` is cancelled (for example because the resource YAML is edited, or the process is shutting down), it waits for all currently running reconciles to finish gracefully.

In order to keep things simple (a timeout can also be problematic, especially for large data where cleanup can take a while), could we maybe just let garbage stay in the staging connector in the case of cancels – and instead clean it up on the next ingestion run? (I.e. also clear the staging location before writing to it.)

Reply:
Thanks for clarifying the issues with `context.Background()`.
There are some issues with cleaning in next ingestion run.
1. The warehouses write to a temporary directory within staging path. The idea is to avoid conflicts in case the same staging path is used by another connector. Snowflake also recommends writing to an empty directory.
2. We can consider storing the temporary path in model state and clear it but it will add extra latency for the next ingestion.

How about we leave the data as it is during cancelation and recommend setting a short data retention on the bucket? IIRC athena has similar constructs as well where user needs to cleanup `output_location` manually.

Reply:
Makes sense and yeah that sounds like a good idea